### PR TITLE
fix tls handshake unexpected timeout.

### PR DESCRIPTION
**Pull request by thank243**
This a temporary solution. Waiting official fixed it.

### CODE REVIEW

File: transport/gun/gun.go [hunk 1]
```diff
@@ -173,6 +174,12 @@ func NewHTTP2Client(dialFn DialFn, tlsConfig *tls.Config) *http2.Transport {
 		}
 
 		cn := tls.Client(pconn, cfg)
+
+		// fix tls handshake not timeout
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		_ = cn.HandshakeContext(ctx)
+
 		if err := cn.Handshake(); err != nil {
```
Comment (icpz) [inline transport/gun/gun.go:183]: `HandshakeContext` will perform the actual handshake, so we should check it's error and remove the unnecessary `Handshake`:

```go
if err := cn.HandshakeContext(ctx); err != nil {
...
}
```

File: transport/gun/gun.go [hunk 2]
```diff
@@ -173,6 +174,12 @@ func NewHTTP2Client(dialFn DialFn, tlsConfig *tls.Config) *http2.Transport {
 		}
 
 		cn := tls.Client(pconn, cfg)
+
+		// fix tls handshake not timeout
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (icpz) [inline transport/gun/gun.go:179]: We can add a constant `DefaultTLSTimeout = 10 * time.Second` in `constant/adapters.go`, and reference it here:
`ctx, cancel := context.WithTimeout(context.Background(), constant.DefaultTLSTimeout)`

**Review by icpz — CHANGES_REQUESTED** [review]

**Comment by thank243** [comment]
add TLS timeout const in constant.

File: transport/gun/gun.go [hunk 3]
```diff
@@ -5,19 +5,20 @@ package gun
 
 import (
 	"bufio"
+	"context"
 	"crypto/tls"
 	"encoding/binary"
 	"errors"
 	"fmt"
+	"github.com/Dreamacro/clash/common/pool"
+	C "github.com/Dreamacro/clash/constant"
```
Comment (icpz) [inline transport/gun/gun.go:14]: Sorry but please don't change the order and empty lines of the imports.

**Review by icpz — CHANGES_REQUESTED** [review]

**Review by icpz — APPROVED** [review]
LGTM

**Comment by Dreamacro** [comment]
Thanks!

**Pull request merged**